src/boot/efi/console.c

```diff
@@ -190,32 +175,42 @@ static EFI_STATUS mode_auto(UINTN *mode) {
          * standard mode, which is provided by the device manufacturer, so it should
          * be a good mode.
          * Note: MaxMode is the number of modes, not the last mode. */
-        if (ST->ConOut->Mode->MaxMode > 2)
-                *mode = 2;
+        if (ST->ConOut->Mode->MaxMode > CONSOLE_MODE_IMPL_FIRST)
+                return CONSOLE_MODE_IMPL_FIRST;
+
         /* Try again with mode different than zero (assume user requests
          * auto mode due to some problem with mode zero). */
-        else if (ST->ConOut->Mode->MaxMode == 2)
-                *mode = 1;
-        /* Else force mode change to zero. */
-        else
-                *mode = 0;
+        if (ST->ConOut->Mode->MaxMode > CONSOLE_MODE_80_50)
+                return CONSOLE_MODE_80_50;
 
-        return change_mode(*mode);
+        return CONSOLE_MODE_80_25;
 }
 
-EFI_STATUS console_set_mode(UINTN *mode, enum console_mode_change_type how) {
-        assert(mode);
+EFI_STATUS console_set_mode(UINT32 mode) {
+        switch (mode) {
+        case CONSOLE_MODE_KEEP:
+                /* If the firmware indicates the current mode is invalid, change it anyway. */
+                if ((UINT32)ST->ConOut->Mode->Mode == CONSOLE_MODE_INVALID)
+                        return change_mode(CONSOLE_MODE_MIN);
+                return EFI_SUCCESS;
 
-        if (how == CONSOLE_MODE_AUTO)
-                return mode_auto(mode);
+        case CONSOLE_MODE_AUTO:
+                return change_mode(get_auto_mode());
 
-        if (how == CONSOLE_MODE_MAX) {
-                /* Note: MaxMode is the number of modes, not the last mode. */
-                if (ST->ConOut->Mode->MaxMode > 0)
-                        *mode = ST->ConOut->Mode->MaxMode-1;
-                else
-                        *mode = 0;
+        case CONSOLE_MODE_AUTO_MAX:
+                mode = MAX(CONSOLE_MODE_MIN, ST->ConOut->Mode->MaxMode - 1);
```

Comment:
It shouldn't. MaxMode is a INT32, so this should return CONSOLE_MODE_MIN, no?

Reply:
ah, ok, thenlet's hope they don't set it to INT32_MIN...

Reply:
If a firmware vendor is that retarded I'd be surprised if the device actually does anything sensible. but if you insist...

Reply:
maybe add an assert() that verifies our assumptions here

Reply:
This code is outdated. Current code relies on 1LL to prevent underflow and change_mode() to clamp the value into valid range. In the unlikely event of MaxMode being being <= 0 we'd end up trying to set it to 0, which should be fine.